libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2164,6 +2164,52 @@ void TypeChecker::typeCheckFunctionGeneralChecks(
 			m_errorReporter.typeError(errorId, paramArgMap[i]->location(), description);
 		}
 	}
+
+	TypePointers const& returnParameterTypes = _functionType->returnParameterTypes();
+	bool isLibraryCall = (_functionType->kind() == FunctionType::Kind::DelegateCall);
+	bool callRequiresABIEncoding =
+		// ABIEncode/ABIDecode calls not included because they should have been already validated
+		// at this point and they have variadic arguments so they need special handling.
+		_functionType->kind() == FunctionType::Kind::DelegateCall ||
+		_functionType->kind() == FunctionType::Kind::External ||
+		_functionType->kind() == FunctionType::Kind::Creation ||
+		_functionType->kind() == FunctionType::Kind::Event;
+
+	if (callRequiresABIEncoding && !experimentalFeatureActive(ExperimentalFeature::ABIEncoderV2))
+	{
+		solAssert(!isVariadic, "");
+		solAssert(parameterTypes.size() == arguments.size(), "");
+		solAssert(!_functionType->isBareCall(), "");
+		solAssert(*_functionCall.annotation().kind == FunctionCallKind::FunctionCall, "");
```

Comment:
I'm tempted to also assert that

```c++
Type const* argumentType = arguments[i]->annotation().type;
solAssert(
	typeSupportedByOldABIEncoder(*argumentType, isLibraryCall) ==
	typeSupportedByOldABIEncoder(*parameterTypes[i], isLibraryCall),
	""
);
```

but I'm not sure if it's guaranteed to always be true. Can I assume that there will never be a type that's compatible with ABI Encoder v1 but is implicitly convertible to another type that is not?

Reply:
Right. I asked because it does matter if the function is variadic (like `abi.encode()`) and at one point I tried to have a unified check here that would cover it too. But yeah, that's not actually a part of the code I submitted in this PR so it's irrelevant here.